### PR TITLE
[RSM - Diagnostics Mode] Add WC_Stripe_Diagnostics_Trace_Store

### PR DESCRIPTION
**Pull request by wjrosa**
Towards RSM-629
Fixes RSM-632

## Changes proposed in this Pull Request

Adds `WC_Stripe_Diagnostics_Trace_Store`: a plain, **file-backed** trace storage class for the upcoming diagnostics recorder. One class file + one test file + loader wiring + changelog.

### Storage layout
- Each trace → one JSON file at `wp-content/uploads/wc-stripe-diagnostics/<sanitized_id>.json`.
- Shared index → `index.json` in the same directory holding the ordered id list.
- An advisory `flock(LOCK_EX)` on a `.index.lock` file in the same directory serializes writers.
- The directory is created lazily on first write via `FilesystemUtil::mkdir_p_not_indexable()`, which drops a `.htaccess` (deny all) and an empty `index.html` to block direct web access. Falls back to `wp_mkdir_p()` on hosts where the helper is missing.

### Why files instead of `wp_options`?
- **Easier to share / hand off.** A trace is a single file you can attach to a support ticket, scp out, or stream from a REST endpoint without re-decoding. Pulling the same data out of `wp_options` means a DB dump or copy/paste.
- **Stays out of database backups.** Traces are short-lived debug data, not application state — no reason for them to round-trip through DB backup/restore.
- **No autoload concerns.** The previous design had to be careful about `autoload=no` on every `update_option`; with files, the question doesn't come up.

### Invariants
- **Trace count cap:** 200. On insert, the oldest trace file is deleted (FIFO) when the index exceeds the cap.
- **Events-per-trace cap:** 200. New events beyond the cap are silently dropped; `append_event()` returns `false`.
- **Serialized size cap:** 100 KB per trace. If appending an event would exceed the cap, the append is rejected.
- **Status machine:** `pending` → `completed` | `abandoned` via `set_status()`. Unknown statuses are rejected.

### Public surface
```php
$store = new WC_Stripe_Diagnostics_Trace_Store();
$store->create( $session_id, $meta );          // bool
$store->append_event( $session_id, $event );   // bool
$store->set_status( $session_id, $status );    // bool
$store->get( $session_id );                    // array|null
$store->delete( $session_id );                 // bool
$store->delete_all();                          // void  (tests + future uninstall)
$store->get_all_ids();                         // string[]
$store->count();                               // int
$store->is_full();                             // bool
$store->get_trace_path( $session_id );         // string|null  (for streaming/sharing)
WC_Stripe_Diagnostics_Trace_Store::sanitize_id( $raw );
```

Session ids are **client-generated**, so `sanitize_id()` strips every non-`[A-Za-z0-9_-]` char and truncates to 64 — no raw id ever becomes part of a filename.

### Concurrency note
Writers acquire `flock(LOCK_EX)` on a dedicated lock file before mutating the index. This is a real OS-level advisory lock (vs. the transient marker the earlier draft used), so concurrent writers serialize cleanly. Reads don't lock; the worst case for an in-flight write is a reader seeing a slightly stale view, not corruption — every per-trace `file_put_contents()` is atomic from the reader's perspective. If `fopen()` on the lock file fails (read-only fs, exhausted descriptors), the critical section still runs unguarded — better than dropping events.

## Testing instructions

No runtime effect yet — no subscribers, no UI, no REST endpoint. This is pure infrastructure. To verify:

1. Run the new PHPUnit test: `npm run test:php -- --filter 'WC_Stripe_Diagnostics_Trace_Store_Test'` — 12 tests pass.
2. Smoke-test in a WP shell (e.g., `wp shell` inside Docker):
   ```php
   $s = new WC_Stripe_Diagnostics_Trace_Store();
   $s->create( 'abc' );
   $s->append_event( 'abc', [ 'kind' => 'test' ] );
   var_dump( $s->get( 'abc' ) );
   echo $s->get_trace_path( 'abc' ); // wp-content/uploads/wc-stripe-diagnostics/abc.json
   ```
3. Inspect `wp-content/uploads/wc-stripe-diagnostics/` and confirm the JSON file is on disk plus a `.htaccess` denying access and an `index.json` listing the trace id.

## Test coverage
- 12 tests covering: round-trip, create idempotency, event append, event cap, size cap, status transitions, FIFO eviction at 200, delete, sanitize_id edge cases, is_full, **`get_trace_path()` returning a real on-disk file**, and `get_trace_path()` returning null for invalid ids.

Ran locally: `npm run lint:php` ✅ · `npm run phpstan` ✅ · targeted PHPUnit (CI)

## Dependencies / stacking
- Base branch: `feature/diagnostics`
- Companion PRs that build on this:
  - Redaction engine (independent, parallel)
  - Recorder singleton (needs this + redaction + #5344)
  - REST events endpoint (needs this — will use `get_trace_path()` for streaming)
  - Action Scheduler cleanup jobs (needs this)

## Changelog

- [x] Covered with tests
- [x] Tested on mobile — N/A (storage class)

Added as `Dev` in `changelog.txt` and `readme.txt`:
> Add WC_Stripe_Diagnostics_Trace_Store (file-backed trace storage with FIFO eviction and size caps) for the upcoming diagnostics recorder

### CODE REVIEW

File: includes/diagnostics/class-wc-stripe-diagnostics-trace-store.php [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+	 *
+	 * @param callable $callback The critical section to run under the lock.
+	 * @return mixed|null
+	 */
+	private function with_lock( callable $callback ) {
+		// Best-effort lock: transients can silently fail under object-cache
+		// contention, and we'd rather accept a rare lost write than drop the
+		// event entirely. The failure mode is a concurrent reader's stale view,
+		// not corruption (each trace row is a single atomic option write).
+		set_transient( self::LOCK_TRANSIENT, 1, self::LOCK_TIMEOUT );
```
[the remark this answers is not in the document]
Comment (wjrosa) [reply includes/diagnostics/class-wc-stripe-diagnostics-trace-store.php]: Dismissing — this comment is obsolete after the file-based refactor (commit 0e70f9c0). The transient "lock" is gone; `with_lock()` now uses `flock(LOCK_EX)` on a dedicated `.index.lock` file, which is a real OS-level advisory lock and provides actual mutual exclusion between writers. The class docstring was updated to match.

File: includes/diagnostics/class-wc-stripe-diagnostics-trace-store.php [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+	/**
+	 * Promote a trace to a terminal status (completed | abandoned).
+	 *
+	 * @param string $session_id Session identifier.
+	 * @param string $status     One of STATUS_COMPLETED, STATUS_ABANDONED.
+	 * @return bool True if status was updated.
+	 */
+	public function set_status( $session_id, $status ) {
+		$session_id = self::sanitize_id( $session_id );
```
[the remark this answers is not in the document]
Comment (wjrosa) [reply includes/diagnostics/class-wc-stripe-diagnostics-trace-store.php:131]: Fixed in 7a4d4461 — moved the empty-id check to immediately follow `sanitize_id()`, before the status validation, matching every other public method.

File: includes/diagnostics/class-wc-stripe-diagnostics-trace-store.php [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+ *
+ * A "trace" is the redacted record of a single checkout session keyed by a
+ * client-generated sessionId. Each trace is persisted as a single JSON file
+ * inside a dedicated, non-indexable directory under the WordPress uploads
+ * folder, alongside an `index.json` file holding the ordered list of trace
+ * ids.
+ *
+ * Storing traces as files (rather than wp_options rows) makes it easy to
+ * download, attach, or hand off an individual trace, and keeps diagnostics
+ * data out of regular database backups.
```
Comment (malithsen) [inline includes/diagnostics/class-wc-stripe-diagnostics-trace-store.php]: super-nit: This paragraph might not be very useful long term. We discussed the pros/cons of using a file vs wp_options, but I'm not sure whether mentioning this here would be useful to future readers.

**Review by malithsen — APPROVED** [review]
This looks great to me. I tested with the snippet to manually invoke the store, and I can see the trace files in `uploads/` 💯 

We can test this more comprehensively and iterate as needed after we wire this up with the rest of the implementation.

File: includes/diagnostics/class-wc-stripe-diagnostics-trace-store.php (hunk 3 above)
Comment (wjrosa) [reply includes/diagnostics/class-wc-stripe-diagnostics-trace-store.php]: Agreed — the file-vs-options rationale is decision context that belongs in the PR discussion, not in the class docblock. A future reader doesn't need to know which alternatives we rejected to understand what the class does. Removed in 4048bdab8.
